### ISSUE TITLE
freeing storage space 

### ISSUE THREAD

**Issue by dvirtz** [ruslo/hunter#1226]
Hunter now takes 50 GB on my hard disk. Some of this is old Hunter versions or old compilers I no longer use. What is the easiest way to free that storage without deleting and rebuilding everything?

**Comment by ruslo** [comment]
Currently there is no any utility that can help with that. Just step into Hunter root and delete what you think is old:
* https://docs.hunter.sh/en/latest/quick-start/uninstall.html

**Comment by isaachier** [comment]
Personally, I just delete `.hunter` periodically when I find it is getting large.

**Comment by dvirtz** [comment]
Thanks. @ruslo do you have any idea for such a utility? I'll be happy to try and contribute to this.
I believe it's a common problem for package managers, maybe there's a solution we can build upon.

**Comment by isaachier** [comment]
I'll look into what maven does seeing as it has a very similar scheme for Java libraries.

**Comment by isaachier** [comment]
As far as I can tell, maven has no support for this. Therefore, not really sure it is worth implementing.

**Comment by ruslo** [comment]
> I believe it's a common problem for package managers

For me it's always a pain :) For example when my Ubuntu Linux machine run out of the disk space I run `dpkg` to see what packages installed and see huge list of everything including dependencies. Meaning that I can't figure out what is installed as while experimenting on something and what is installed for day-by-day usage. Then I remove some stuff, run `autoremove` and check disk space again. Repeat until satisfied. Pretty same with `brew` on OSX.

For `apt-get` and `brew` it's better because update doesn't create new directory. Effectively they have only one `Hunter-ID/Toolchain-ID/Config-ID` triple always. It's worse for reproducibility/customization but better for user experience.

> do you have any idea for such a utility?

Need to discuss. From my understanding the best behavior is to decide what amount of disk space is allowed to be used by Hunter and let Hunter remove old directories when limit reached. It means that it should be implemented in Hunter and customization done by user somewhere in `~/.config/Hunter` (see https://docs.hunter.sh/en/latest/user-guides/cmake-user/protected-sources.html). But this leave a lot of implementation notes. For example there is no instrument in CMake that allow to calculate size of file or directory. But the trickiest part will be to decide what exactly to remove (https://en.wikipedia.org/wiki/Cache_replacement_policies) based on information how often root directory used, what was the last time it was used, does build directory that trigger install exists (I guess should not be used by CI system and make sense only for user's local machine), etc.

For now as a workaround it will be easier to create external script but you need to be sure that there is no Hunter related builds are running because Hunter is not designed to work correctly when somebody removing internal stuff (two Hunter instances synchronized with each other while sharing same resources). If you violate this you can leave Hunter root directory in invalid state which user will not be able to escape, this include cache hence probably with invalid uploads on server.

**Comment by OvermindDL1** [comment]
At least a platform dependent clean script in the `.hunter` directory would be nice to purge the 50% oldest things or something.

**Comment by ruslo** [comment]
> to purge the 50% oldest things or something

What is the "50% oldest things" exactly? Back to:

> But the trickiest part will be to decide what exactly to remove (https://en.wikipedia.org/wiki/Cache_replacement_policies) based on information ...

**Comment by OvermindDL1** [comment]
Could make it configurable, but honestly being lazy and just deleting anything that is older than, say, a month is fine (or let it take an argument like `5d` for anything older than 5 days, or `0` for all or so, or perhaps take something like `1gb` so it purges things from the oldest on until 1gb remains).  The user ran the clean so they know they might end up recompiling some dependency on their next recompile anyway.

Still, I'd probably keep just wiping it all clean, and probably add it to a cron job, though with a script like that I might keep the latest 5gb or 1M(onth) or so.  It doesn't need to be fancy, it is just a helper script since the user can always go through the directories manually anyway (I've ran `du ~/.hunter -h` a few times to see what suddenly is eating 20 gigs... could even add such options like it into the clean script).

**Comment by ruslo** [comment]
Probably it's better to start without "old vs new" feature and remove general directories:
```
Downloads/
Cache/
Cellar/
<Hunter-ID>/
```

`Downloads` is directory with source archives. Sources + build = `Cache`. `Cache` unpacked to `Cellar`, links from `Cellar` goes to `<Hunter-ID>/`.

Few types of cleaning:
* `all`: remove everything: `Downloads`, `Cache`, `Cellar`, `<Hunter-ID>`
  * user have to downloads sources and start all builds from scratch
* `cache`: remove `Cache`, `Cellar` and all `<Hunter-ID>`, keep `Downloads`
  * no need to download sources, start build from scratch
* `cellar`: remove `Cellar` and `<Hunter-ID>`, keep `Downloads` and `Cache`. 
  * no need to build, just unpack binaries from `Cache`
* `install`: remove `<Hunter-ID>`, keep `Cellar`, `Cache` and `Downloads`. Probably make no sense since it's just links and will not affect disk usage

**Comment by OvermindDL1** [comment]
Though I have to admit, it would be nice to clean it all, run my builds, tell it to take a snapshot of what currently exist, and just clean anything that does not match the snapshot.  :-)

**Comment by ruslo** [comment]
> deleting anything that is older than, say, a month is fine

That's the problem, it may be not what user expect in practice. Say you have stable stuff with few gigobytes of files and want to try small example with Qt and OpenCV. You install Qt/OpenCV, play with it and removed. Now you're starting new build, space exhausted and Hunter have to remove something. Instead of removing unused new Qt/OpenCV it removes stable old.

**Comment by ruslo** [comment]
> Few types of cleaning

May be reword it to:
* `clean=all` remove everything. User have to download sources and run build again
* `clean=cached` remove everything except downloads. User have to run build again
* `clean=compact` remove only `Cellar` and roots `<Hunter-ID>`. User only have to wait for unpack/link

Plus `--show-disk-usage` to control what effect each step had.

**Comment by OvermindDL1** [comment]
> That's the problem, it may be not what user expect in practice. Say you have stable stuff with few gigabytes of files and want to try small example with Qt and OpenCV. You install Qt/OpenCV, play with it and removed. Now you're starting new build, space exhausted and Hunter have to remove something. Instead of removing unused new Qt/OpenCV it removes stable old.

Yep, that is why I think snapshots would be nice.  The default snapshot would just be empty so doing something like `clean=nonsnapshotted` or so would just wipe everything anyway.  I'd love to run that in a cronjob once a week or so once I get a snapshot built.  :-)

**Comment by dvirtz** [comment]
I have a different idea. Would it be possible for each package to keep
references to CMake projects using it and then go over all packages and if
all references are either deleted or upgraded Hunter version or toolchain
then remove the package?

**Comment by ruslo** [comment]
> keep references to CMake projects using it

Basically it's what I meant by

> But the trickiest part will be to decide what exactly to remove ... does build directory that trigger install exists (I guess should not be used by CI system and make sense only for user's local machine), etc.

Need to introduce `USED` stamps with information about build directory. Will raise a concern about out-of-date `USED` stamps since you can update your CMake code with new HunterGate URL/SHA1.

**Comment by caseymcc** [comment]
On a side note, somewhat related. What about setting up hunter gate to use the latest hunter on the machine (assuming this does not already exist), rather than spelling out the version. I know my hunter folder generally gets large because each project I am messing with is setup for different version of hunter. (and running through changing them all to the same version is a bit of a pain).

If a project needs a specific version they can could choose it the way we currently do with the header declaration, otherwise let it check for the latest one in the .hunter folder and use that.

**Comment by ruslo** [comment]
> setting up hunter gate to use the latest hunter on the machine

How do you know that?

> I know my hunter folder generally gets large because each project I am messing with is setup for different version of hunter

You can use environment variable `HUNTER_DISABLE_AUTOINSTALL`:
* https://docs.hunter.sh/en/latest/reference/user-variables.html#hunter-disable-autoinstall

> otherwise let it check for the latest one (or fetch the latest if none exists) in the .hunter folder and use that

Actually I guess it's possible. Read something from `~/.config/Hunter/*.cmake`?

**Comment by wheybags** [comment]
Please don't do that :s Reproducible builds is one of the best things about
hunter

On Nov 28, 2017 11:57 PM, "Ruslan Baratov" <notifications@github.com> wrote:

setting up hunter gate to use the latest hunter on the machine

How do you know that?

I know my hunter folder generally gets large because each project I am
messing with is setup for different version of hunter

You can use environment variable HUNTER_DISABLE_AUTOINSTALL:

   - https://docs.hunter.sh/en/latest/reference/user-
   variables.html#hunter-disable-autoinstall
   <https://docs.hunter.sh/en/latest/reference/user-variables.html#hunter-disable-autoinstall>

otherwise let it check for the latest one (or fetch the latest if none
exists) in the .hunter folder and use that

Actually I guess it's possible. Read something from ~/.config/Hunter/*.cmake
?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/ruslo/hunter/issues/1226#issuecomment-347705230>, or mute
the thread

**Comment by caseymcc** [comment]
I think as long as you keep the ability to set the hunter version for a project then you still have reproducible builds, which is what is currently required by default. I see this working just like not calling out a project version, say if I want to use opencv I can just hunter_add_package(opencv) and hunter will use the latest listed in the defaults, and everytime I upgrade hunter version opencv might be updated as well. The option is there if need to call out an opencv version but not required, just like we could for hunter itself.

**Comment by ruslo** [comment]
> Please don't do that :s Reproducible builds is one of the best things about hunter

Reproducibility is one of the main features and there is no plans to change it. The talk about optional feature that user can enable if it fits his needs.

@caseymcc Please create separate issue for further discussion of global HunterGate parameters if you want to add something more/implementation details. Let's focus on disk usage here.
